can-value.js

```diff
@@ -1 +1,80 @@
-module.exports = {};
+var canKey = require("can-key");
+var canKeyUtils = require("can-key/utils");
+var canReflect = require("can-reflect");
+var canReflectDependencies = require("can-reflect-dependencies");
+var keyObservable = require("can-simple-observable/key/key");
+var Observation = require("can-observation");
+
+module.exports = {
+	bind: function(object, keyPath) {
+		return keyObservable(object, keyPath);
+	},
+
+	from: function(object, keyPath) {
+		var observation = new Observation(function() {
+			return canKey.get(object, keyPath);
+		});
+
+		canReflect.assignSymbols(observation, {
+			"can.getName": function getName() {
+				var objectName = canReflect.getName(object);
+				return "CanValueFromObservation<" + objectName + "." + keyPath + ">";
+			}
+		});
+
+		return observation;
+	},
+
+	to: function(object, keyPath) {
+		var keyPathParts = canKeyUtils.parts(keyPath);
+		var lastIndex = keyPathParts.length - 1;
+		var lastKey;// This stores the last part of the keyPath, e.g. “key” in “outer.inner.key”
+		var lastParent;// This stores the object that the last key is on, e.g. “outer.inner” in outer: {inner: {"key": "value"}}
+		var observable = {};
+
+		// Walk the object path to set up the dependency/mutation data
+		var previousKeyData;
+		canKey.walk(object, keyPathParts, function(keyData, i) {
+			if (i === lastIndex) {
+				// observable is mutating keyData.parent
+				lastKey = keyData.key;
+				lastParent = keyData.parent;
+				canReflectDependencies.addMutatedBy(lastParent, lastKey, observable);
+
+				canReflect.onKeyValue(previousKeyData.parent, previousKeyData.key, function(newValue, oldValue) {
+					canReflectDependencies.deleteMutatedBy(oldValue, lastKey, observable);
+					canReflectDependencies.addMutatedBy(newValue, lastKey, observable);
+					lastParent = newValue;
+				});
```

Comment:
Where there is an `onKeyValue` there must be an `offKeyValue`… I don’t think this logic is correct.

In short, if `outer.inner` changes, I need to make sure the dependency data for`outer.inner.key` is updated (old one deleted, new one added).

I think this means I should also be looking for _any_ changes in the keypath, so hypothetically if the entire path were `wrapper.outer.inner.key`, then I would need to listen to `wrapper.outer` being set too, so I could get the new `wrapper.outer.inner` and set up its dependency data.

Reply:
Talked to Justin about this today; some changes I’m going to make to the `to()` method:

- Use `keyObservable()` and remove `can.getValue` symbol from the returned instance
- Add `can.onValue` and `can.offValue` symbols that will create an observation similar to what’s in `from()` to keep track of the objects along the keypath
- Wrap all the dependency/mutation data and debug names so they’re removed in production
- Name the Observation returned by `from()` (instead of the `can.getName` on it now)